### ISSUE TITLE
Check that 16S ASVs are correct with barrnap

### ISSUE THREAD

**Issue by erikrikarddaniel** [nf-core/ampliseq#317]
## Is your feature request related to a problem? Please describe

Some ASV sequences are not what we think they are. They might be products of primers binding to other parts of genomes, or just sequencing errors.

## Describe the solution you'd like

Using [barrnap](https://github.com/tseemann/barrnap) we could identify SSU rRNA sequences and optionally throw away other sequences. This step would ideally come right after chimera removal but needs to be optional.

For other sequences I can't see any simple solutions, nor can I see that we can easily generalize barrnap to other hmm models.

**Comment by d4straub** [comment]
Sounds great, just had a project where there was mostly non-16S amplicons, very annoying. But those had also not the complete primers in their sequences, only 60-80% of the primer sequence and were thrown out by cutadapt with current standard settings. Still, some remained even after that step, would be good to filter them out like this.

Just one question: Is there a specifc advantage/disadvantage to use barrnap after ASV computation instead of before?

**Comment by erikrikarddaniel** [comment]
> Sounds great, just had a project where there was mostly non-16S amplicons, very annoying. But those had also not the complete primers in their sequences, only 60-80% of the primer sequence and were thrown out by cutadapt with current standard settings. Still, some remained even after that step, would be good to filter them out like this.
> 
> Just one question: Is there a specifc advantage/disadvantage to use barrnap after ASV computation instead of before?

I think the longer the sequence the better, plus working on the raw sequences would take much longer. 

It will affect counts however, which is of course a bit of a problem as one would end up with ASV tables before and after barrnap -- unless one overwrites -- so I'm willing to reconsider.

**Comment by d4straub** [comment]
I just played a little and got for input file `ASV_seqs.fasta`:
```
singularity pull https://depot.galaxyproject.org/singularity/barrnap:0.9--hdfd78af_4
singularity shell barrnap:0.9--hdfd78af_4

total=$(grep -c "^>" ASV_seqs.fasta)
for KINGDOM in bac arc mito euk
do
    barrnap --threads 2 --quiet --reject 0.1 --incseq --kingdom $KINGDOM --outseq ASV_seqs.filtered.${KINGDOM}.fasta < ASV_seqs.fasta > ASV_seqs.rrna.${KINGDOM}.gff
    filtered=$(grep -c "^>" ASV_seqs.filtered.${KINGDOM}.fasta)
    percent=$(awk -v var1="$filtered" -v var2="$total" 'BEGIN { print  ( var1 / var2 *100 ) }')
    echo "Found $percent % in kingdom $KINGDOM"
done
```
After testing, I see a few problems, referring to the output above:
1. `ASV_seqs.filtered.${KINGDOM}.fasta` contains all sequences that passed the filter, but they are trimmed. We do not want to have trimmed sequences I think, but sequences that matched.
2. `ASV_seqs.rrna.${KINGDOM}.gff` contains in combination with `--incseq` all (also non-matching) complete sequences, also not what we need.
3. from 1 and 2 follows that we would need to extract all ASV md5sums that matched (`ASV_seqs.filtered.${KINGDOM}.fasta`) and filter `ASV_seqs.fasta` with it. Thats annoying.
4. I chose `--reject 0.1` so that only 10% match of the full length rRNA sequence will make the sequence pass. This is fine for V4 region (14-15% match) but obviously that number should be increased with longer ASVs. So that threshold should be best computed by looking into sequence length of the input `ASV_seqs.fasta` and checking the reference input length. Best would be to have an option for the minimum absolute length, not %!
5. Optimally, we would also let people choose what `--kingdom` they want to filter for, default could be all, as we have here. I.e. a new parameter that allows a comma-separated list, probably `--asv_filter_kingdom "bac,arc"`.

Those issues seem all solvable, but not nice, would you agree?

**Comment by erikrikarddaniel** [comment]
> I just played a little and got for input file `ASV_seqs.fasta`:
> 
> ```
> singularity pull https://depot.galaxyproject.org/singularity/barrnap:0.9--hdfd78af_4
> singularity shell barrnap:0.9--hdfd78af_4
> 
> total=$(grep -c "^>" ASV_seqs.fasta)
> for KINGDOM in bac arc mito euk
> do
>     barrnap --threads 2 --quiet --reject 0.1 --incseq --kingdom $KINGDOM --outseq ASV_seqs.filtered.${KINGDOM}.fasta < ASV_seqs.fasta > ASV_seqs.rrna.${KINGDOM}.gff
>     filtered=$(grep -c "^>" ASV_seqs.filtered.${KINGDOM}.fasta)
>     percent=$(awk -v var1="$filtered" -v var2="$total" 'BEGIN { print  ( var1 / var2 *100 ) }')
>     echo "Found $percent % in kingdom $KINGDOM"
> done
> ```
> 
> After testing, I see a few problems, referring to the output above:
> 
>     1. `ASV_seqs.filtered.${KINGDOM}.fasta` contains all sequences that passed the filter, but they are trimmed. We do not want to have trimmed sequences I think, but sequences that matched.
> 
Completely agree. I suppose one can pull out the ids/names from the fasta and put into a tsv. By the end, it would be good with a five column tsv: id plus the four different "domains" that barrnap supports, i.e. a boolean (match or not) or probability or whatever.

>     2. `ASV_seqs.rrna.${KINGDOM}.gff` contains in combination with `--incseq` all (also non-matching) complete sequences, also not what we need.
> 
>     3. from 1 and 2 follows that we would need to extract all ASV md5sums that matched (`ASV_seqs.filtered.${KINGDOM}.fasta`) and filter `ASV_seqs.fasta` with it. Thats annoying.
> 
I think the tsv I propose above would be easier (for me at least). Easier to pull out of the barrnap results too.

>     4. I chose `--reject 0.1` so that only 10% match of the full length rRNA sequence will make the sequence pass. This is fine for V4 region (14-15% match) but obviously that number should be increased with longer ASVs. So that threshold should be best computed by looking into sequence length of the input `ASV_seqs.fasta` and checking the reference input length. Best would be to have an option for the minimum absolute length, not %!
> 
I think I agree. Can't really see all the consequences now.

>     5. Optimally, we would also let people choose what `--kingdom` they want to filter for, default could be all, as we have here. I.e. a new parameter that allows a comma-separated list, probably `--asv_filter_kingdom "bac,arc"`.
> 
See above. I see no reason not to run all "kingdoms" for all projects.

> Those issues seem all solvable, but not nice, would you agree?

Absolutely! Thanks for looking into this.

**Comment by d4straub** [comment]
It might be actually much easier to just use [sortmerna](https://github.com/biocore/sortmerna) that has also already an [nf-core module](https://github.com/nf-core/modules/blob/master/modules/sortmerna/main.nf) which is currently not putting the rRNA reads into an output channel (so much for generalisation) but that could be changed. First need to test sortmerna though if it indeed produces immediately what we want.

**Comment by erikrikarddaniel** [comment]
> It might be actually much easier to just use [sortmerna](https://github.com/biocore/sortmerna) that has also already an [nf-core module](https://github.com/nf-core/modules/blob/master/modules/sortmerna/main.nf) which is currently not putting the rRNA reads into an output channel (so much for generalisation) but that could be changed. First need to test sortmerna though if it indeed produces immediately what we want.

Could be a possibility. I think the only reason barrnap exists is because the Prokka author had issues with the license for sortmerna. If it fit us better, why not?

**Comment by d4straub** [comment]
I could not find any reference to sortmerna, only to RNAmmer at https://github.com/tseemann/barrnap#comparison-with-rnammer:
> Barrnap is designed to be a substitute for [RNAmmer](http://www.cbs.dtu.dk/services/RNAmmer/). 

Sortmerna has the exact same licensing as Barrnap (GPLv3), as far as I can see.

**Comment by erikrikarddaniel** [comment]
Right, I had confused sortmerna with RNAmmer.

Another thing: A colleague got suspicious when I mentioned we were thinking of using another tool than barrnap since that's what we used for the SBDI-GTDB taxonomy database. I'm not convinced it matters.

What I do when I run barrnap, to get a tsv, is that I first run the tool with all four types of SSU profiles, then:

```
grep -h '>' *.fna | sed 's/^>//' | sed 's/:\+/\t/g' | sort -u > output.tsv
```

In case one would like to do this, I suppose that calls for two modules, one for barrnap called four times, and one to collect the output tsv.

**Comment by d4straub** [comment]
Hi, I added in the above PR barrnap via the parameter `--filter_ssu` which takes arguments for the "kingdoms" of barrnap. 
- Is there any other setting we would really need a parameter for? (a config with `-c` can always modify it as well)
- When using `--filter_ssu` the sequences & ASV table will be filtered, is it required to be able to use barrnap but not filter the data? (I would say no, in a special case one would need to run the pipeline twice)

**Comment by erikrikarddaniel** [comment]
Can `--filter_ssu` take a list?
Since barrnap is relatively cheap to run, wouldn't it be a good idea to run all four domains by default and output a table? If the user specifies `--filter_ssu` the table could be used to filte.
(Sorry for not coming up with that idea earlier...)

**Comment by d4straub** [comment]
Yes, `--filter_ssu` can take e.g. "bac" or a comma separated list, e.g. "bac,arc,mito". But if you think that this most likely isnt import, than I can use that parameter as a boolean whether ASVs are filtered or not and run barrnap on default on all 4 kingdoms.

**Comment by erikrikarddaniel** [comment]
That's not what I meant. The flag is perfect as it is. But in some cases a user might be interested in a table but not in filtering (I think we are), so my suggestion was to always run barrnap for all four and filter when the parameter is passed.

**Comment by d4straub** [comment]
Ok, I am uncertain whether I understand right or if I am misinterpreting. 

As far as I understand, this would be the desired scenario:
- run barrnap by default on "bac,arc,mito,euk"
- by default, do not filter based on barrnap results, enable filter on only request

So my solution for this could be:
- `--barrnap_kingdom`, default `"bac,arc,mito,euk"`, controls kindoms, takes a comma separated list of kingdoms to test
- `--filter_ssu` as boolean, default `false`, enables ASV filtering based on barrnap results

Does that sound good?

**Comment by erikrikarddaniel** [comment]
Yes, although I would actually prefer to be able to run barrnap on all four, then filter on only, say bacteria and archaea making both params lists. This might be a very rare use case, I don't know. The reason is that I think there will be overlap between the domains, so that a sequence matches more than one profile sometimes, which one might want to look into. (In most cases, I will probably run barrnap on all four, but not use the filtering, instead filtering myself in post-processing, so no big deal for me.)

**Comment by d4straub** [comment]
> Yes, although I would actually prefer to be able to run barrnap on all four, then filter on only, say bacteria and archaea making both params lists. This might be a very rare use case, I don't know. 

Ah, I see. That is also possible... than no `--barrnap_kingdom` is required but `--filter_ssu` would take the kingdom. I'll think about it!

> The reason is that I think there will be overlap between the domains, so that a sequence matches more than one profile sometimes, which one might want to look into. 

Yes, I have seen with the test dataset that there are relatively large overlaps. But evalues seems to differ also alot.

> (In most cases, I will probably run barrnap on all four, but not use the filtering, instead filtering myself in post-processing, so no big deal for me.)

Alright, and what files would you prefer for this? In the current PR, the .gff's for each kingdom are copied into the results folder, are those sufficient? After all they contain all info (evalues, match ranges, etc). Additionally, one file that only contains unique matches (that last file might change/be removed when I adapt the filtering as above).

**Comment by erikrikarddaniel** [comment]
> > Yes, although I would actually prefer to be able to run barrnap on all four, then filter on only, say bacteria and archaea making both params lists. This might be a very rare use case, I don't know.
> 
> Ah, I see. That is also possible... I'll think about it! I might make it simpler and omit `--barrnap_kingdom` (i.e. barrnap runs always on all kingdoms and exports separate matches per kingdom) but `--filter_ssu` would take the kingdom for the subsequent filtering.
> 
> > The reason is that I think there will be overlap between the domains, so that a sequence matches more than one profile sometimes, which one might want to look into.
> 
> Yes, I have seen with the test dataset that there are relatively large overlaps. But evalues seems to differ also alot.
> 
> > (In most cases, I will probably run barrnap on all four, but not use the filtering, instead filtering myself in post-processing, so no big deal for me.)
> 
> Alright, and what files would you prefer for this? In the current PR, the .gff's for each kingdom are copied into the results folder, are those sufficient? After all they contain all info (evalues, match ranges, etc). Additionally, one file that only contains unique matches (that last file might change/be removed when I adapt the filtering as above).

That sounds pretty good. gffs are, afterall tsvs. Having them all in one file is practical of course.

**Comment by d4straub** [comment]
Ok, changes as discussed are added. Documentation & edge case handling (0 matches) have to be improved.
